fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/TempTableInsertPlan.java

```diff
@@ -52,84 +60,135 @@
-        super(inner, null, (Type.Record)((Type.Relation)tempTableReferenceValue.getResultType()).getInnerType(), null,
-                null, computationValue, currentModifiedRecordAlias());
+                                @Nonnull final Value tempTableReferenceValue,
+                                boolean isOwningTempTable) {
+        this.inner = inner;
         this.tempTableReferenceValue = tempTableReferenceValue;
+        this.isOwningTempTable = isOwningTempTable;
     }
 
+    @Nonnull
     @Override
-    public PipelineOperation getPipelineOperation() {
-        return PipelineOperation.INSERT;
+    @SuppressWarnings({"unchecked"})
+    public <M extends Message> RecordCursor<QueryResult> executePlan(@Nonnull final FDBRecordStoreBase<M> store,
+                                                                     @Nonnull final EvaluationContext context,
+                                                                     @Nullable final byte[] continuation,
+                                                                     @Nonnull final ExecuteProperties executeProperties) {
+        if (isOwningTempTable) {
+            final var typeDescriptor = getInnerTypeDescriptor();
+            return TempTableInsertCursor.from(continuation,
+                    proto -> {
+                        final var tempTable = Objects.requireNonNull((TempTable<QueryResult>)getTempTableReferenceValue().eval(store, context));
+                        if (proto != null) {
+                            TempTable.from(proto, typeDescriptor).getIterator().forEachRemaining(tempTable::add);
+                        }
+                        return tempTable;
+                    },
+                    childContinuation -> getChild().executePlan(store, context, childContinuation, executeProperties.clearSkipAndLimit()));
+        } else {
+            return getChild().executePlan(store, context, continuation, executeProperties.clearSkipAndLimit())
+                    .map(queryResult ->
+                    {
+                        final var tempTable = Objects.requireNonNull((TempTable<QueryResult>)getTempTableReferenceValue().eval(store, context));
+                        tempTable.add(queryResult);
+                        return queryResult;
+                    });
+        }
     }
 
     @Nullable
-    @Override
-    protected <M extends Message> Descriptors.Descriptor getTargetDescriptor(@Nonnull final FDBRecordStoreBase<M> ignored) {
-        return null;
+    private Descriptors.Descriptor getInnerTypeDescriptor() {
+        final Descriptors.Descriptor typeDescriptor;
+        if (tempTableReferenceValue.getResultType().isRelation() && ((Type.Relation)tempTableReferenceValue.getResultType()).getInnerType().isRecord()) {
+            final var type = (Type.Record)((Type.Relation)tempTableReferenceValue.getResultType()).getInnerType();
+            final var builder = TypeRepository.newBuilder();
```

Comment:
Done. I think we don't need to pass the type information as a forth parameter in the factory method, since it must be equal to `((Type.Relation)inner.getRangesOver().getResultType()).getInnerType()` considering that we don't support value promotions and other transformations in the current version.

https://github.com/FoundationDB/fdb-record-layer/blob/124ccceabbd54478afa3557d6b21361c0e9f5d67/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/expressions/TempTableInsertExpression.java#L188-L191

Reply:
that is correct I think